### ISSUE TITLE
kairos cluster v3.0.0 not created when setting e.g. MTUBytes=1450 in /etc/systemd/network/  .network file

### ISSUE THREAD

**Issue by Wolverine65** [kairos-io/kairos#2358]
**Kairos version:**
```
PRETTY_NAME="Ubuntu 22.04.4 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.4 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
KAIROS_NAME="kairos-standard-ubuntu-22.04"
KAIROS_MODEL="generic"
KAIROS_HOME_URL="https://github.com/kairos-io/kairos"
KAIROS_SOFTWARE_VERSION="v1.29.0+k3s1"
KAIROS_VERSION="v3.0.0-v1.29.0-k3s1"
KAIROS_IMAGE_LABEL="22.04-standard-amd64-generic-v3.0.0-k3sv1.29.0-k3s1"
KAIROS_FLAVOR_RELEASE="22.04"
KAIROS_VARIANT="standard"
KAIROS_REGISTRY_AND_ORG="quay.io/kairos"
KAIROS_BUG_REPORT_URL="https://github.com/kairos-io/kairos/issues"
KAIROS_PRETTY_NAME="kairos-standard-ubuntu-22.04 v3.0.0-v1.29.0-k3s1"
KAIROS_FLAVOR="ubuntu"
KAIROS_VERSION_ID="v3.0.0-v1.29.0-k3s1"
KAIROS_IMAGE_REPO="quay.io/kairos/ubuntu:22.04-standard-amd64-generic-v3.0.0-k3sv1.29.0-k3s1"
KAIROS_ARTIFACT="kairos-ubuntu-22.04-standard-amd64-generic-v3.0.0-k3sv1.29.0+k3s1"
KAIROS_TARGETARCH="amd64"
KAIROS_RELEASE="v3.0.0"
KAIROS_GITHUB_REPO="kairos-io/kairos"
KAIROS_ID="kairos"
KAIROS_ID_LIKE="kairos-standard-ubuntu-22.04"
KAIROS_SOFTWARE_VERSION_PREFIX="k3s"
```

**CPU architecture, OS, and Version:**
Linux 3-kairosdev-sandbox-2d4bd1gx 6.5.0-25-generic #25~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Feb 20 16:09:15 UTC 2 x86_64 x86_64 x86_64 GNU/Linux

**Describe the bug**
We use a non default mtusize in our network stack.
Since kairos version >2.4.1  setting MTUBytes=1450 for network interfaces a cluster wil not be created
Creating e.g. /etc/systemd/networkd/05-setmtu.network with MTUBytes=1450 results in nodes not 'finding' each other to setup cluster
```
            [Match]
            Name=ens1*
            [Network]
            DHCP=no
            [Link]
            MTUBytes=1450
            [DHCP]
            ClientIdentifier=mac
```
When MTUBytes=1500 ( default ) the  cluster is created, manually setting MTUBytes=1450 in /oem/90_custom.yaml after the cluster has been created and a reboot, cluster is still functioning.

**To Reproduce**
Create /etc/systemd/network/05-ens1.network with MTUBytes=1450 set within cloud-init config and deploy nodes
```
[Match]
Name=ens1*
[Network]
DHCP=yes
[Link]
MTUBytes=1450
[DHCP]
ClientIdentifier=mac
```

**Expected behavior**

Cluster should be created with non default MTUBytes set

**Logs**
```
{"level":"info","ts":1710768172.4428058,"caller":"service/node.go:308","msg":"Applying role 'auto'"}
{"level":"info","ts":1710768172.4428377,"caller":"service/role.go:115","msg":"Role loaded. Applying auto"}
{"level":"info","ts":1710768172.4439757,"caller":"role/auto.go:31","msg":"Active nodes:[12D3KooWD1ivphPNr8gasxAR4FUM88NCRaqbfvbonanxJAiH3uv5]"}
{"level":"info","ts":1710768172.444072,"caller":"role/auto.go:32","msg":"Advertizing nodes:[cc186c5488c430b39ab27a7a65f81a27-3-kairosdev-sandbox-2d4bd1gx]"}
{"level":"info","ts":1710768172.4440954,"caller":"role/auto.go:35","msg":"Not enough nodes"}
{"level":"info","ts":1710768172.4441204,"caller":"service/node.go:357","msg":"not enough nodes available, sleeping... needed: 2, available: 1"}
```
**Additional context**
none

**Comment by Itxaka** [comment]
Hey @Wolverine65 

after the file is created in the cloud config, have you tried to do a `networkctl reload` ?

Can you post the cloud config used for the deployment of the cluster (redacting anything not needed)?

**Comment by Wolverine65** [comment]
Hello @Itxaka ,

I also tried a networkctl reload, even added that to the runcmd part of the cloud-init. Unfortunately no luck. Below is the config we use to deploy for v2.4.1 which works fine for that version

```
#cloud-config

install:
  # Device for automated installs
  device: auto
  # Reboot after installation
  reboot: false
  # Power off after installation
  poweroff: false
  # Set to true to enable automated installations
  auto: true

# The upgrade block configures what happens when upgrade is called
upgrade:
  # Reboot after upgrade
  reboot: true
  # Power off after upgrade
  poweroff: true

growpart:
  devices: ['/']

stages:
  boot:
    - name: "Set system keymap"
      system_firstboot:
        keymap: us
    - name: "Setup hostname"
      hostname: "{{ vm_hostname }}"
    - name: "Set up various kube environment variables"
      environment:
        KUBECONFIG: /etc/rancher/k3s/k3s.yaml
        CONTAINERD_ADDRESS: /run/k3s/containerd/containerd.sock
        CONTAINERD_NAMESPACE: k8s.io
  
  initramfs:
    - files:
        - path: /etc/systemd/network/01-manual.network
          permissions: 0644
          content: |
            [Match]
            Name=ens1*
            [Network]
            DHCP=yes
            [Link]
            MTUBytes = 1450
        - path: /etc/systemd/network/02-manual.network
          permissions: 0644
          content: |
            [Match]
            Name=ens2*
            [Network]
            DHCP=no
            [Link]
            MTUBytes = 1450

users:
- name: "redacted"                             # The username for the user.
  passwd: "redacted"                      # The password for the user.
  lock_passwd: true
  groups:
  - "admin"
  ssh_authorized_keys:                      # A list of SSH keys to add to the user's authorized keys.
{% for pubkeys in ssh_keys %}
  - {{ pubkeys }}
{% endfor %}

write_files:
  - path: /etc/sysctl.conf
    content: |
{% for parameters in sysctlconf %}
         {{ parameters }}
{% endfor %}
runcmd:
  - /usr/sbin/sysctl -p

k3s:
  enabled: true
  args:
  - --default-local-storage-path=/opt/k3s-local-storage
  - --disable='traefik,servicelb'
  - --disable-cloud-controller
  - --flannel-backend=none
  - --disable-network-policy
  - --cluster-cidr="{{ lookup('vars', cluster_dictionary)[cluster_name].cluster_cidr }}"
  - --service-cidr="{{ lookup('vars', cluster_dictionary)[cluster_name].service_cidr }}"
  - --tls-san "{{ lookup('vars', cluster_dictionary)[cluster_name].kube_vip[otap] }}"
  - --node-taint 'node-role.kubernetes.io/master=true:NoSchedule'
  # following standard kairos
  - --cluster-dns="{{ lookup('vars', cluster_dictionary)[cluster_name].cluster_dns }}"
  - --cluster-domain="{{ lookup('vars', cluster_dictionary)[cluster_name].cluster_domain }}"
  - --kube-controller-manager-arg 'bind-address=0.0.0.0'
  - --kube-scheduler-arg 'bind-address=0.0.0.0'
  - --kube-controller-manager-arg 'terminated-pod-gc-threshold=10'
  env:
    K3S_TOKEN: "{{ lookup('vars', cluster_dictionary)[cluster_name].k3s_token[otap] }}"
{% if lookup('vars', cluster_dictionary)[cluster_name].ha[otap] == false %}
    # create database external
    K3S_DATASTORE_ENDPOINT: "mysql://{{ lookup('vars', cluster_dictionary)[cluster_name].mysql[otap].dbuser }}:{{ lookup('vars', cluster_dictionary)[cluster_name].mysql[otap].dbpassword }}@tcp({{ lookup('vars', cluster_dictionary)[cluster_name].mysql[otap].dbhost }}:{{ lookup('vars', cluster_dictionary)[cluster_name].mysql_port }})/{{ lookup('vars', cluster_dictionary)[cluster_name].mysql[otap].dbname }}"
{% endif %}

# before installation
bundles:
- targets:
  - run://quay.io/kairos/packages:k9s-utils-0.26.7

# create network token
# docker run -ti --rm quay.io/mudler/edgevpn -b -g
p2p:
  disable_dht: true
  network_token: "{{ lookup('vars', cluster_dictionary)[cluster_name].network_token[otap] }}"
  vpn:
    # Disable VPN, so traffic is not configured with a VPN
    create: false
    use: false
{% if lookup('vars', cluster_dictionary)[cluster_name].ha[otap] == true %}
  auto:
    enable: true
    ha:
      enable: true
      master_nodes: {{ lookup('vars', cluster_dictionary)[cluster_name].master_nodes[otap] | default ('2') }}
      external_db: "mysql://{{ lookup('vars', cluster_dictionary)[cluster_name].mysql[otap].dbuser }}:{{ lookup('vars', cluster_dictionary)[cluster_name].mysql[otap].dbpassword }}@tcp({{ lookup('vars', cluster_dictionary)[cluster_name].mysql[otap].dbhost }}:{{ lookup('vars', cluster_dictionary)[cluster_name].mysql_port }})/{{ lookup('vars', cluster_dictionary)[cluster_name].mysql[otap].dbname }}"
{% endif %}

# Choose free ip from netbeheer, and use us alsi in the k3s tls-san ( see above)
kubevip:
  eip: "{{ lookup('vars', cluster_dictionary)[cluster_name].kube_vip[otap] }}"
  enable: true
  interface: "ens192"
  manifest_url: "https://kube-vip.io/manifests/rbac.yaml"
  # manifest_url: "https://raw.githubusercontent.com/kube-vip/kube-vip-cloud-provider/main/manifest/kube-vip-cloud-controller.yaml"


```

**Comment by Itxaka** [comment]
> Below is the config we use to deploy for v2.4.1 which works fine for that version

ahh this is very important, thanks for pointing out. I'll have a quick look tomorrow to see if its something obvious and has a quick fix.

**Comment by jimmykarily** [comment]
Possibly relevant link: https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/1850704

**Comment by jimmykarily** [comment]
I'm confused by the options used in the provided example config.

- `p2p.auto.enable: true` 
- `p2p.network_token` (set)
- `p2p.vpn.create: false`
- `p2p.vpn.use: false`
 
What is this combination supposed to do?
I'm trying to reproduce the problem with a minimum configuration.
 I have to slim it down removing the templated values and anything that is not relevant. But I'm not sure what the intended setup is supposed to be. For example, although it seems that this is a non-automated setup (no p2p etc),
still I don't see any k3s args pointing the agent to the master node (K3S_URL ?).

Can you please clarify what the intended setup is? We should come up with minimum reproducing steps that will show that simply changing the MTU value make it fail, otherwise it could easily be some other option missing or being wrong.

**Comment by Wolverine65** [comment]
hello @jimmykarily ,

this is an automated setup
we use the same cloud-init  config for all our nodes. The kairos-agent/p2p then decides which one is a master and which are the workers

```
p2p:
  disable_dht: true
  network_token: "< TOKEN >"
  vpn:
    create: false
    use: false
  auto:
    enable: true
    ha:
      enable: true
      master_nodes: 2
      external_db: "< MYSQL STRING >"
```

The problem is when we try to set a non default mtu size for our network interfaces, the p2p functionality doesnt seem to find the nodes

            [Link]
            MTUBytes = 1450

When i change it to 1500 , it works, but we need a size of 1450

This has worked up until v2.4.1 , so something has changed.

Perhaps it is related to https://github.com/kairos-io/kairos/issues/2103  , since we also filed that issue, but not sure about that

Regards,
H

**Comment by Wolverine65** [comment]
also tested with debian, same result.
set mtusize=1450 through dhcp options from dhcp server, also same result peer nodes dont find each other.

**Comment by ci-robbot** [comment]
It seems the issue already existed before ([see issue 2088](https://github.com/kairos-io/kairos-net/issues/2088)), and it appears to be related to how edgevpn handles MTU settings.
https://github.com/mudler/edgevpn/issues/90

I'll open another issue to track the problem/discussion about this topic related to the kairos version: https://github.com/kairos-io/kairos-net/issues/2110

**Comment by Wolverine65** [comment]
Hello, 

any update on this issue? 

Thanks.

**Comment by ci-robbot** [comment]
Refreshed triage on `master` (kairos-io/kairos `9946553f`), two years after my last note. Still open, and the shape of the fix is now clear enough to pick.

## The MTU defaults have not moved

`mudler/edgevpn` pins at `v0.35.3` in the monorepo `go.mod`. The flag defaults are byte-identical to `v0.23.1`:

- `--mtu` (interface MTU on the `edgevpn0` tun): **1200**
- `--packet-mtu` (per-frame payload sent between peers over libp2p): **1420**

`cmd/util.go` at both `v0.23.1:60-90` and current HEAD sets these same values, and `pkg/vpn/interface.go`'s `LinkSetMTU(link, c.InterfaceMTU)` is unchanged. So the regression between kairos `v2.4.1` and later is not in edgevpn's MTU code.

## What kairos passes to the edgevpn daemon

`provider/internal/provider/p2p.go:163` (`SetupVPN`) writes `/etc/systemd/system.conf.d/edgevpn-kairos.env` and never sets `EDGEVPNMTU` or `EDGEVPNPACKETMTU`. Users get the edgevpn defaults unless they override them via `p2p.vpn.env` (a straight `map[string]string` that `applyAPIListenerEnv` merges into the daemon env at `p2p.go:136`).

## Why physical MTU 1450 breaks with the defaults

edgevpn frames a payload of `PacketMTU` bytes (1420) and hands it to libp2p, which wraps it in TCP or QUIC + libp2p framing + IP + link headers. On a 1500-byte link there is roughly 60-80 bytes of slack. On the reporter's 1450-byte physical link, a 1420-byte payload plus overhead lands over the wire size, so peers can't complete the discovery/handshake in either direction. That matches "nodes don't find each other" without an obvious log signature.

## Workaround (verified path, not verified in a repro)

Lower `EDGEVPNPACKETMTU` below the physical MTU minus overhead in cloud-config, so the daemon reads it from the systemd env file:

```yaml
p2p:
  vpn:
    env:
      EDGEVPNPACKETMTU: "1300"
```

1300 leaves ~150 bytes of headroom on a 1450 link, which covers libp2p + QUIC/TCP + IP + link comfortably.

## Where the real regression sits

edgevpn's `pkg/node/connection.go` uses `libp2p.DefaultTransports` / `libp2p.DefaultPrivateTransports`. Between `v0.23.1` and current, `go-libp2p` moved from `v0.27.1` to `v0.48.0`, and the shape of what "default transports" means changed materially in that window (QUIC-v1 stabilized, WebTransport landed, resource manager arrived, `holepunch`/`identify` framing evolved). Any one of those changes can shift the effective packet size the transport puts on the wire, which is why the same 1420 payload that used to fit through a 1450 link no longer does.

## Proposed fix, smallest wedge first

Three options, ordered by effort:

1. **Docs.** Add the `p2p.vpn.env.EDGEVPNPACKETMTU` recipe to the p2p troubleshooting section on kairos-docs. Cheapest, unblocks users today.
2. **Auto-lower in kairos-provider.** In `SetupVPN`, read the MTU of the default-route interface at write time; if it is < 1500, set `EDGEVPNPACKETMTU` to `MTU - 150` unless the user already set it in `p2p.vpn.env`. Contained to `provider/internal/provider/p2p.go`, one probe helper.
3. **Upstream libp2p bisect.** Trace which `go-libp2p` bump between `v0.27.1` and `v0.48.0` shifted the effective on-wire MTU, and either pin/patch edgevpn or teach it to compute `PacketMTU` from the link. Real work, upstream.

I would do (1) plus (2) in this repo and leave (3) to a dedicated edgevpn issue. Happy to open a docs PR now and a provider PR after, but I want a maintainer nod before touching the auto-lower path — auto-tuning MTU is exactly the kind of change that surprises someone in a way they cannot easily undo.

**Comment by ci-robbot** [comment]
Docs PR is up: kairos-io/kairos-docs#685.

Only the workaround from option (1) in the triage above. Adds a
Troubleshooting section to the p2p page with the
`p2p.vpn.env.EDGEVPNPACKETMTU` recipe and the sizing rule
(physical MTU minus ~150 bytes for libp2p + transport + IP + link
overhead). No code changes and the same value has to land on every
node.

Still leaving the provider-side auto-lower for a maintainer nod, per
the note above.

**Comment by ci-robbot** [comment]
Correction to my 2026-09-04 triage here: the `EDGEVPNPACKETMTU` workaround I
proposed does not work, and I have moved the docs PR (kairos-io/kairos-docs#685)
back to draft.

I reproduced the scenario on `edgevpn v0.35.5`, the pin Kairos ships, with two
netns on a veth pair at clamped MTUs. Three findings, full logs and captures at
https://gist.github.com/ci-robbot/77ffc392eb73946f0b805458f5607b8c:

1. **The premise does not reproduce at 1450.** With stock config, peers
   converged in 3-4s at link MTU 1450, 1400 and 1300, and passed 4000-byte
   fragmented pings. So `MTUBytes=1450` alone is not what broke this cluster.
2. **`EDGEVPNPACKETMTU` cannot bind with the shipped defaults.** `getFrame` in
   `pkg/vpn/vpn.go` uses it to size the tun *read buffer* and returns
   `frame[:n]`; nothing pads to it. The default tun MTU is 1200, so any value
   above 1200 is a no-op, and setting it low enough to bind truncates real
   traffic (at `EDGEVPNMTU=1400 EDGEVPNPACKETMTU=1300`, 1300-byte pings were
   100% loss). `EDGEVPNMTU` is the knob that bounds payload size, not
   `EDGEVPNPACKETMTU`.
3. **The real cliff is at 1200, not 1500.** Link MTU 1200, 1100 and 1000 all
   time out after 90s with raw L3 confirmed good, and neither knob rescues
   them, nor does forcing a TCP-only transport. I also modelled a broken-PMTUD
   tunnel (asymmetric 1500/1450, oversized frames dropped with no ICMP
   frag-needed) and defaults still converged.

So this issue is still open and still unexplained. The next step is finding out
what the reporter's link actually did, since it is not the packet size. Worth
noting the 1200 cliff looks like a separate, reproducible limitation.

**Comment by ci-robbot** [comment]
Follow-up to my 2026-09-16 correction. I could not rebuild the netns harness this session (namespaces are blocked in my current sandbox), so I went at the transport layer instead with a root-free probe, and it gives a hard number for one half of this.

**Kairos p2p loses the QUIC transport entirely on any IPv4 link with an MTU below 1308 bytes, silently.**

### Measurement

A size-limiting UDP relay on the path, dropping any datagram above a limit in either direction. That models a link whose MTU caps the UDP payload without needing root. Against `quic-go v0.60.0`, the pin `go-libp2p v0.48.0` and `edgevpn v0.35.x` resolve to:

```
## default config (InitialPacketSize unset -> protocol.InitialPacketSize = 1280)
udp_limit  result   dropped_up
1280       OK       0
1279       FAIL     10           context deadline exceeded
1252       FAIL     10
1228       FAIL     10
1200       FAIL     10
```

The floor is exactly 1280, and it is a cliff, not a degradation. `dropped_up 10` is every Initial retransmit being discarded: quic-go pads Initials to `protocol.InitialPacketSize` and **never probes downward**, so it retries the same oversized datagram until the handshake times out.

Adding IPv4 overhead (20 + 8), that is a **link MTU of 1308**. For IPv6 (40 + 8) it is 1328.

### Why it is invisible from Kairos

`protocol.InitialPacketSize = 1280` is a quic-go constant, but `quic.Config.InitialPacketSize` can lower it to the RFC 9000 floor of 1200. Same probe with that set:

```
## InitialPacketSize = 1200
1200       OK       0
1199       FAIL     10
```

So the reachable floor is 1228 (IPv4), and below that QUIC is impossible by protocol. But `go-libp2p` does not expose or set it: `p2p/transport/quicreuse/config.go` builds its `quic.Config` with eight fields and `InitialPacketSize` is not one of them, so the 1280 default stands. Nothing in Kairos, provider-kairos or edgevpn can reach it either. A fix would be upstream in go-libp2p.

Harness and full sweeps: https://gist.github.com/ci-robbot/10a7e040ad91ae817498e07635204064

### What this does and does not explain

It does explain the shape of what I measured on 2026-09-16: link MTU 1300 converged, and 1300 < 1308, so QUIC was already dead there and the TCP transport carried it. It also means every sub-1308 deployment has been running p2p on TCP alone without any signal saying so.

It does **not** explain the 1200 cliff, and I want to be clear about that rather than tidy it away. TCP has no comparable floor, so TCP should have carried 1200, 1100 and 1000 as well, and it did not. Either the second failure has a different cause, or my TCP-only run (`EDGEVPNLISTENMADDRS=/ip4/0.0.0.0/tcp/0`) did not actually take QUIC out of the dial path. Re-running the netns sweep with per-transport dial logs is the next step, and it needs an environment where I can create namespaces.

This still leaves the reporter's original case unexplained: 1450 is above 1308, so QUIC was available on @Wolverine65's link and the packet size is not what broke it.
